Zenmode Optional Header and Menubar

soamaven opened the pull request:
I tweaked the Zenmode extension to be able to show the header and menubar. I was getting annoyed from having to exit Zenmode everytime I wanted to access the menubar, so I thought I'd make it an option. Did the same for header because it was easy.

Also, from everything I can tell this extension is 5.x compatible.

jcb91 commented:
Mainly looks reasonable, but should it also affect the unhiding behaviour? Perhaps just restoring them to their pre-zen state? Perhaps I'm overthinking this though...

soamaven commented:
I don't understand. All the js should do is toggle the `hidden` flag. Can you elaborate on what you mean affect the unhiding behaviour and restoring their pre-zen state? This is to keep the nice background of zen while having easy access to the menubar (for things like 'close and halt')

jcb91 commented:
Well, as it currently stands, the modified version now only *hides* the menubar/header as set by the config. However, when zen is toggled off again, it will *unhide* both again (circa line 73), regardless of what state they started in. This could be annoying if, for example, one wished zen to just not affect the menubar at all, and set that independently. Perhaps this level of detail is unimportant though, since in that case I'd have to have been prepared to manually hide/unhide the menubar anyway!

jcb91 commented:
Anyway, I'm happy to merge as-is, but you'll need a quick release/merge to handle the recent change to config loading 😄

soamaven commented:
>However, when zen is toggled off again, it will unhide both again (circa line 73), regardless of what state they started in.

I understand what you mean. That could get annoying if I had header hidden, went zen, and then back out of zen to find the header. I unfortunately don't know how to fix this, but this is how Zenmode operates currently anyway, so it's not decreasing functionality.

The new config loading breaks the functionality... I am not sure what's going on. Before I just mimicked the config loading of the other options, and it worked fine.

jcb91 commented:
> this is how Zenmode operates currently anyway, so it's not decreasing functionality

True, I was being overly picky, this can wait till somebody reports being inconvenienced by it.

> The new config loading breaks the functionality... I am not sure what's going on. Before I just mimicked the config loading of the other options, and it worked fine.

Ah... do you mean that the new loading breaks *all* functionality, or just the bit you've added? Could you push your merged/rebased branch so I can take a look?

soamaven commented:
> or just the bit you've added?

Yeah just the added bits

> Could you push your merged/rebased branch so I can take a look?

Done.

soamaven commented:
>So it looks like rather than merging, you've essentially just gone back to the version before the new config loading mechanism - see individual comments for further details. I can't see an obvious reason why the new behaviour would be broken, but let's get the merge sorted first before we drill down into that!

Sorry. Self taught here. Git is easy to use if you stay in the shallows, but the drop-off is steep. I understand the request now. 

So I rebased onto the latest commit. I added in the options, and those can be seen in the configurator. This happens due to the yaml file being correct, so that's not surprising.

The main.js though, doesn't work. It doesn't show the zenmode button on the toolbar. Where can i find the debug log? Reinstalling the extension with `--debug` isn't giving anything to the console. Starting up with `jupyter notebook --debug` doesn't give much information either.

jcb91 commented on src/jupyter_contrib_nbextensions/nbextensions/zenmode/main.js:
```diff
@@ -13,12 +13,10 @@ define([
     require,
     $,
     IPython,
-    events
```
The events object is needed in new line 107, so this message to go back in. Everything else looks good :)

jcb91 commented:
> Sorry. Self taught here. Git is easy to use if you stay in the shallows, but the drop-off is steep. I understand the request now.

No worries! Definitely, my apologies, I probably should have provided a better explanation of what I meant!

> Also, where can i find the debug log? Reinstalling the extension with `--debug` isn't giving anything to the console. Starting up with `jupyter notebook --debug` doesn't give much information either. I imagine this would have saved me some time.

So, for the JavaScript nbextensions, any useful logs will be in the browser's
[javascript console](http://webmasters.stackexchange.com/questions/8525/how-to-open-the-javascript-console-in-different-browsers). Sometimes you can also get useful information from the server logs when running `jupyter notebook --debug`, such as which is & CSS files are being requested by the browser.

Thanks for your persistence and patience on this! 😄

jcb91 commented on src/jupyter_contrib_nbextensions/nbextensions/zenmode/main.js:
```diff
@@ -13,20 +13,28 @@ define([
-    "base/js/events"
+    "base/js/events",
+    'base/js/utils',
+    'services/config',
```
Apologies, when I said "None of this hunk is needed", I meant "none of the _changes_ in this hunk", we still need the events module 😅

jcb91 commented:
@soamaven apologies for the delay looking at this, but anyway, it looks good now! 👍
